rules/DeadCode/Rector/Foreach_/RemoveUnusedForeachKeyRector.php

```diff
@@ -48,31 +51,51 @@ public function getRuleDefinition(): RuleDefinition
      */
     public function getNodeTypes(): array
     {
-        return [Foreach_::class];
+        return [StmtsAwareInterface::class];
     }
 
     /**
-     * @param Foreach_ $node
+     * @param StmtsAwareInterface $node
      */
     public function refactor(Node $node): ?Node
     {
-        if (! $node->keyVar instanceof Expr) {
+        if ($node->stmts === null) {
             return null;
         }
 
-        $keyVar = $node->keyVar;
+        $hasChanged = false;
+        foreach ($node->stmts as $key => $stmt) {
+            if (! $stmt instanceof Foreach_) {
+                continue;
+            }
 
-        $isNodeUsed = (bool) $this->betterNodeFinder->findFirst(
-            $node->stmts,
-            fn (Node $node): bool => $this->nodeComparator->areNodesEqual($node, $keyVar)
-        );
+            if (! $stmt->keyVar instanceof Variable) {
+                continue;
+            }
 
-        if ($isNodeUsed) {
-            return null;
+            $keyVar = $stmt->keyVar;
+
+            $isNodeUsed = (bool) $this->betterNodeFinder->findFirst(
+                $stmt->stmts,
+                fn (Node $node): bool => $this->nodeComparator->areNodesEqual($node, $keyVar)
+            );
+
+            if ($isNodeUsed) {
+                continue;
+            }
+
+            if ($this->stmtsManipulator->isVariableUsedInNextStmt($node, $key + 1, (string) $this->getName($keyVar))) {
```

Comment:
does this mean, the logic only works when the var usage is directly after the loop?

the initial problem also exists on e.g.

```
<?php

namespace Rector\Tests\DeadCode\Rector\Foreach_\RemoveUnusedForeachKeyRector\Fixture;

final class SkipKeyUsedInNexStmt
{
    function lastKey ($container, $defKey = null) {
        $result = $defKey;

        foreach ($container as $result => $value) {
          // next
        }

       $x = 1;
       $y = 2;

        return $result;
      }
}
```

Reply:
The logic is check all next stmts:

```
        $stmts = array_slice($stmtsAware->stmts, $jumpToKey, null, true);
        return (bool) $this->betterNodeFinder->findVariableOfName($stmts, $variableName);
```